articles/realtime.md

```diff
@@ -0,0 +1,259 @@
+
+In general, C++ patterns with poor cache locality are not well-suited to real-time environments.
+Another such pattern is the opaque pointer idiom (PIMPL), which is convenient for ABI compatibility and speeding up compile times.
+However, bouncing between the memory location for the object and its private data pointer causes the cache to "spill" as it loads one chunk of memory and then another, unrelated chunk for almost every function in the PIMPLized object.
+
+### Exceptions
+
+Throwing an exception can put large objects on the stack, which is undesirable in real-time programming since we cannot allocate memory on the heap.
+On modern C++ compilers, catching an exception has little memory or time overhead, but can lead to unexpected code growth.
+
+## Device I/O
+Interacting with physical devices (disk I/O, printing to the screen, etc.) may introduce unacceptable latency in the real-time code path, since the process is often forced to wait on slow physical phenomena.
+Additionally, many I/O calls such as `fopen` result in pagefaults.
+
+Keep disk reads/writes at the beginning or end of the program, outside of the RT code path.
+
+Spin up userspace threads (executing on a different CPU from the RT code) to print output to the screen.
+
+## Multithreaded Programming and Synchronization
+
+Real-time computation requirements change the typical paradigm of multithreaded programming.
+Program execution may not block asynchronously, and threads must be scheduled deterministically.
+A real-time operating system will fulfill this scheduling requirement, but there are still pitfalls for the developer to fall into.
+This section provides guidelines for avoiding these pitfalls.
+
+### Thread creation guidelines
+
+Create threads at the start of the program.
+This confines the nondeterministic overhead of thread allocation to a defined point in the process.
+
+Create high priority (but not 99) threads with a FIFO or Round Robin scheduler.
+
+### Avoid priority inversion
+Priority inversion can occur on a system with a preemptive task scheduler and results in deadlock.
+It occurs when: a low-priority task acquires a lock and is then pre-empted by a medium-priority task, then a high-priority task acquires the lock held by the low-priority task.
+
+The three tasks are stuck in a triangle: the high-priority task is blocked on the low-priority task, which is blocked on the medium-priority task because it was preempted by a task with a higher priority, and the medium-priority task is also blocked on a task with a higher priority.
+
+Here are the some solutions to priority inversion:
+
+* Don't use locks (but sometimes, they are necessary)
+* Disable preemption for tasks holding locks (can lead to jitter)
+* Increase priority of task holding a lock
+* Use priority inheritance: a task that owns a lock inherits the priority of a task that tries to acquire the lock
+
```

Comment:
Another alternative: Use lock-free data structures. They don't have blocking semantics and guarantee forward progress of the active thread. On the flipside, their performance is typically poorer than locks, and can have a larger memory footprint (because of multiple versioned objects being tracked). A [very nice read on the subject](http://www.1024cores.net/home/lock-free-algorithms).